### ISSUE TITLE
mounted apps can method dispatch

### ISSUE THREAD

**Issue by faassen** [morepath/morepath#490]
When an application is mounted using a new instance of the app is returned from the mount directive. The first time you call a dispatch method on a new app instance, it's not bound to the instance yet. With the mount system, each time you traverse into a mounted application, the app instance is new. This means that ``__get__`` is triggered, which isn't welcome.

An application developer could go around this by caching application instances, but that's something they didn't need to do before, and it's really an implementation detail. Ideally we'd do the binding earlier, preferably during the configuration process, but we don't have the instances yet there.

**Comment by faassen** [comment]
One way to solve this and solve #479 would be to require developers to implement ```__hash__``` for their application classes. This way application instancing can be cached. This would be okay, as applications are treated immutably. But it would require additional work on the part of the developers.

Another way to get this hashing information would be to extend the mount directive to ask for it: a ``get_app_key`` argument. If it isn't there, generate it based on the arguments to the application object.

**Comment by faassen** [comment]
Morepath should also use the application class as a hash key, but it can do this automatically. I'm starting to lean towards the ``get_app_key`` solution.

**Comment by faassen** [comment]
Perhaps ``get_app_key`` should have three modes: explicitly given by the developer, like:

```
lambda app: (app.name, app.blah)
```

auto generated from the arguments for the ``App`` class. Or completely absent, in which case the existing behavior is retained, including the bug for #479.

**Comment by taschini** [comment]
> This means that`` __get__`` is triggered, which isn't welcome.

Well, it isn't dramatic either. The slow path in ``__get__``, whereby a new instance of ``DispatchMethod`` is created, is taken only if the internal cache of the descriptor is not hit.

**Comment by faassen** [comment]
That's true, though it does hinder debugging to run into ``__get__`` at all. For a moment I thought the first invocation hook would also be triggered but that's incorrect, as ``obj`` is not ``None``.
